istio/discovery.go

```diff
@@ -0,0 +1,561 @@
+	if revision != "default" && revision != "" {
+		configMapName = configMapName + "-" + revision
+	}
+
+	return configMapName
+}
+
+// Discovery detects istio infrastructure and configuration across clusters.
+type Discovery struct {
+	conf           *config.Config
+	kialiCache     cache.KialiCache
+	kialiSAClients map[string]kubernetes.ClientInterface
+}
+
+// NewDiscovery initializes a new Discovery.
+func NewDiscovery(kialiSAClients map[string]kubernetes.ClientInterface, cache cache.KialiCache, conf *config.Config) *Discovery {
+	return &Discovery{
+		conf:           conf,
+		kialiCache:     cache,
+		kialiSAClients: kialiSAClients,
+	}
+}
+
+// IsRemoteCluster determines if the cluster has a controlplane or if it's a remote cluster without one.
+// Clusters that do not exist or are not accessible are considered remote clusters.
+func (in *Discovery) IsRemoteCluster(ctx context.Context, cluster string) bool {
+	mesh, err := in.Mesh(ctx)
+	if err != nil {
+		log.Debugf("Unable to get mesh to determine if cluster [%s] is remote. Err: %s", cluster, err)
+		return false
+	}
+
+	// If there's a controlplane for the cluster then it's not a remote cluster.
+	for _, controlPlane := range mesh.ControlPlanes {
+		if controlPlane.Cluster.Name == cluster {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Clusters resolves the Kubernetes clusters that are hosting the mesh. Resolution
+// is done as best-effort using the resources that are present in the cluster.
+// TODO: should this go in kubernetes package?
+func (in *Discovery) Clusters() ([]models.KubeCluster, error) {
+	if clusters := in.kialiCache.GetClusters(); clusters != nil {
+		return clusters, nil
+	}
+
+	// Even if somehow there are no clusters found, which there should always be at least the homecluster,
+	// setting this to an empty slice will prevent us from trying to resolve again.
+	clustersByName := map[string]models.KubeCluster{}
+	for cluster, client := range in.kialiSAClients {
+		info := client.ClusterInfo()
+		meshCluster := models.KubeCluster{
+			// If there's a client for this cluster then it's accessible.
+			Accessible: true,
+			Name:       cluster,
+			SecretName: info.SecretName,
+		}
+		if info.ClientConfig != nil {
+			meshCluster.ApiEndpoint = info.ClientConfig.Host
+		}
+
+		meshCluster.Network = in.resolveNetwork(cluster)
+
+		if cluster == in.conf.KubernetesConfig.ClusterName {
+			meshCluster.IsKialiHome = true
+		}
+		clustersByName[cluster] = meshCluster
+	}
+
+	// Add clusters from config.
+	for _, cluster := range in.conf.Clustering.Clusters {
+		if _, found := clustersByName[cluster.Name]; !found {
+			clustersByName[cluster.Name] = models.KubeCluster{
+				Name: cluster.Name,
+				// Clusters without a SecretName are not accessible
+				// because we don't have a kubeconfig for them.
+				Accessible: cluster.SecretName != "",
+			}
+		}
+	}
+
+	clusters := maps.Values(clustersByName)
```

Comment:
That shouldn't ever happen since the clusters are derived from the kiali service account clients and kiali wouldn't startup if there were no service accounts. I can add some logging just in case though.